### PR TITLE
rework torcx uploading to include a manifest

### PR DESCRIPTION
**Pull request by euank**
- [x] `./build_packages` produces manifest and cas pkgs
- [x] `./build_packages --upload` does the right thing for developers
- [x] `./build_image` on dev machines includes torcx packages based on a manifest correctly
- [x] `./jenkins/images.sh` correctly populates torcx store and can `build_image`
- [x] `kola` accepts a 'torcx manifest' path to test a manifest prior to signing / upload (https://github.com/coreos/mantle/pull/701)
- [ ] signing scripts for download/signing/uploading the torcx manifest

### CODE REVIEW

**Comment by euank** [comment]
It might be helpful to have an example manifest handy for the sake of reviewing, especially the torcx_manifest library helpers.

Here's what `./build_torcx_store --upload` spits out currently:
https://storage.googleapis.com/users.developer.core-os.net/esk/torcx/manifests/amd64-usr/1520.0.0%2B2017-09-06-1330/torcx_manifest.json

File: build_torcx_store [hunk 1]
```diff
@@ -151,10 +167,52 @@ function torcx_package() {
...
+                update_default=true
+                pkg_locations+=("/usr/share/torcx/store/${name}:${version}.torcx.tgz")
+        fi
+	if [[ "${FLAGS_upload}" -eq ${FLAGS_TRUE} ]]; then
+		pkg_locations+=("$(download_tectonic_torcx_url "pkgs/${BOARD}/${name}/${digest}/${name}:${version}.torcx.tgz")")
+	fi
+        torcx_manifest::add_pkg "${manifest_path}" \
+            "${name}" \
+            "${version}" \
+            "sha512-${sha512sum}" \
```
Comment (squeed) [inline build_torcx_store]: Could the separator be a ':' ? That matches the opencontainers digest format.

Comment (lucab) [reply build_torcx_store]: I think the answer is no, as this matches ignition format instead.

Comment (euank) [reply build_torcx_store]: Yeah, I'd prefer the arbitrary format ignition chose over the arbitrary format tarsum chose.

File: build_torcx_store [hunk 2]
```diff
@@ -170,14 +170,20 @@ function torcx_package() {
...
+        # Ideally we should move to something more like a casync digest or tarsum.
+        # The reason this is currently not being done is because to do that we
+        # *MUST* ensure that a given pair of (digest, sha512sum) referenced in
+        # a previous torcx package remains correct.
+        # Because this code, as written, clobbers existing things with the same
+        # digest (but the sha512sum of the .torcx.tgz can differ, e.g. due to ctime)
+        # that property doesn't hold.
+        # To switch this back to a reprodicble digest, we *must* never clobber
+        # existing objects (and thus re-use their sha512sum here).
+        digest="${sha512sum}"
```
Comment (dm0-) [inline build_torcx_store]: Is `$sha512sum` defined at this point?

Comment (euank) [reply build_torcx_store]: Nope; this is broken because of that. Patch incoming once it actually runs.

Comment (euank) [reply build_torcx_store]: .. Should be fixed now.

File: build_library/release_util.sh [hunk 3]
```diff
@@ -48,6 +56,14 @@ check_gsutil_opts() {
         UPLOAD_ROOT="${FLAGS_upload_root%%/}"
     fi
 
+    if [[ -n "${FLAGS_torcx_upload_root}" ]]; then
+        if [[ "${FLAGS_torcx_upload_root}" != gs://* ]]; then
+            die_notrace "--torcx-upload_root must be a gs:// URL"
```
Comment (dm0-) [inline build_library/release_util.sh]: The option is `--torcx_upload_root`.

File: build_library/torcx_manifest.sh [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+# get_versions returns the list of versions for a given package. 
+function torcx_manifest::get_versions() {
+  local file="${1}"
+  local name="${2}"
+  jq -r ".value.packages[] | select(.name == \"${name}\").versions[].version" < "${file}"
+}
+
+# get_versions returns the default version for a given package, or an empty string if there isn't one.
```
Comment (dm0-) [inline build_library/torcx_manifest.sh]: Wrong function name.

File: build_torcx_store [hunk 5]
```diff
@@ -166,13 +221,44 @@ DEFAULT_IMAGES=(
         =app-torcx/docker-17.06
 )
 
+# Tihs list contains extra images which will be uploaded and included in the
```
Comment (dm0-) [inline build_torcx_store]: `This`

**Comment by euank** [comment]
Fixed those typoes and squashed down to a more reasonable number of commits

**Comment by squeed** [comment]
🎉

**Pull request merged**